Correct issue #1959 and correct issues in SPIN package

julient31 opened the pull request:
**Summary**

This PR contains corrections for the SPIN package and for the minimization norms.
- magnetic energy agreeing with total energy
- corrected pair/spin interactions to match NVE spin-lattice calculation
- added two validation problems (NVE and NVT spin-lattice calculations)
- corrected the min_modify doc page
- corrected the "inf" norm in min.cpp

**Related Issues**

- Correct issues in the SPIN package (reported by users)
- Correct issue #1959 (min_modify doc page equations)

**Author(s)**

Julien Tranchida (jtranch@sandia.gov)

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

NR

**Implementation Notes**

All examples and validation problems of the SPIN package were checked.

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [x] A package specific README file has been included or updated
- [x] One or more example input decks are included

julient31 commented:
@akohlmey Hey Axel! I included the min_modify doc page correction in the PR. 
Also, I noticed there was an issue with the 'inf' norm in min.cpp (I was not taking the square of the force components when computing its MAX, whereas the convergence criterion is converted into a squared value in min.cpp). Please feel free to have a look. 
Most of the other changes are corrections/improvements in the SPIN package. 
Thanks!

junghans commented on src/SPIN/compute_spin.cpp:
```diff
@@ -217,8 +217,8 @@ void ComputeSpin::compute_vector()
-  // spintemperature /= (2.0*kb*tempdenomtot);
-  spintemperature /= (kb*tempdenomtot);
+  spintemperature /= (2.0*kb*tempdenomtot);
+  // spintemperature /= (kb*tempdenomtot);
```
Could you please remove the commented out line? If anybody ever reads this code, it is really confusing.

julient31 replied:
Sure.

julient31 commented:
@athomps Thanks for checking the PR Aidan.
No I don't think so. If you look at one of the minimization routines (for example, line 118 of min_cg.cpp), the square of the tolerance criterion value is being compared to fdotf. 
To be consistent, the fnorm_inf() routine also needs to compute a squared value.

athomps commented:
These two lines appear to be in conflict:

  fnorm2_init = sqrt(fnorm_sqr());
  fnorminf_init = fnorm_inf();

The first one takes the square root, the second does not.  Also, in finish.cpp, the output defines fnorminf to be:

"  Force max component initial, final"

which means that you need to take the square root.

julient31 commented:
Yes, sorry. I misunderstood your first comment.

julient31 commented:
> These two lines appear to be in conflict:
> 
> fnorm2_init = sqrt(fnorm_sqr());
> fnorminf_init = fnorm_inf();

They were indeed three instances of fnorm_inf() in min.cpp that I missed.

akohlmey commented:
@julient31 please note, that there are also the min_kokkos.cpp and min_cg_kokkos.cpp files that may need updates when changes to the corresponding non-KOKKOS versions are made. It should be straightforward to run tests with KOKKOS for OpenMP binaries compiled via CMake.

```
cd /path/to/lammps
mkdir build-kokkos
cd build-kokkos
cmake3 -C ../cmake/presets/minimal.cmake -C ../cmake/presets/kokkos-openmp.cmake -D PKG_SPIN=on ../cmake
cmake3 --build .
cd /path/to/spin/min/example
/path/to/lammps/build-kokkos/lmp -in in.some-input -k on t 2 -sf kk
```
You should get the same results as for the non-KOKKOS version.

akohlmey commented:
FYI, I just merged in all the recent changes to upstream which include the CMake preset files for KOKKOS.

julient31 commented:
Hey @akohlmey , I modified min_kokkos.cpp and min_cg_kokkos.cpp to match the changes of min.cpp and min_cg.cpp. Although it was fairly straightforward, I would suggest you have a look at it, as I do not have a lot of experience programming in Kokkos. 
I was able to build Lammps with cmake to run Kokkos tests, following your instructions. Everything seemed to be working fine. 
Finally, I updated all the log files of the SPIN examples, accounting for the latest changes.

julient31 commented:
Hey @stanmoore1 and @akohlmey , actually, I'm seeing a small difference. 
I wanted to check the lattice cg minimizer (not the one in the SPIN package) just to make sure. So I used the input script in examples/min. cmake and make builds actually give me slightly different trajectories, even before the minimization starts (during the NVE steps). Does that make sense to you? I'm guessing there might be slight differences in the way the LJ potential is computed in Kokkos. Let me know what you think. 
![make_cmake](https://user-images.githubusercontent.com/28453668/79279525-72274c80-7e6b-11ea-8732-17109d611152.png)

akohlmey commented:
> Hey @stanmoore1 and @akohlmey , actually, I'm seeing a small difference.
> I wanted to check the lattice cg minimizer (not the one in the SPIN package) just to make sure. So I used the input script in examples/min. cmake and make builds actually give me slightly different trajectories, even before the minimization starts (during the NVE steps). Does that make sense to you? Let me know what you think.

yes. the CMake builds now use `-ffast-math -march=native` in addition to the normal compiler flags, which runs significantly faster, but relaxes the stricter floating-point math rules and thus this kind of gradual exponential divergence is happening. A bit more than what you would like to see, but ...

> PS: the make build is _make yes-spin yes-kokkos_ and then _make serial_. the cmake build follows Axel's instructions above.
